### PR TITLE
feat(frontend): implementation for create/drop sink handlers

### PR DESCRIPTION
**Pull request by nanderstabel**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Add handle for CREATE SINK and DROP SINK statements as well as test_utils. 

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/frontend/src/catalog/sink_catalog.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// limitations under the License.
+
+use risingwave_pb::catalog::Sink as ProstSink;
+
+use super::SinkId;
+
+/// this struct `SinkCatalog` is used in frontend and compared with `ProstSink` it only maintains
+/// information which will be used during optimization.
+#[derive(Clone, Debug)]
+pub struct SinkCatalog {
```
Comment (neverchanje) [inline src/frontend/src/catalog/sink_catalog.rs:22]: We will not support querying on sinks in the short term, at least not in the mvp. So pls remove this file for now.

File: src/frontend/src/test_utils.rs [hunk 2]
```diff
@@ -310,6 +338,13 @@ impl MockCatalogWriter {
         Ok(source.id)
     }
 
+    fn create_sink_inner(&self, mut sink: ProstSink) -> Result<u32> {
```
Comment (neverchanje) [inline src/frontend/src/test_utils.rs:341]:
```suggestion
    fn create_sink_inner(&self, mut sink: ProstSink) -> Result<()> {
```

Seriously, if we don't use the id, why do we return it?

File: src/meta/src/rpc/service/ddl_service.rs [hunk 3]
```diff
@@ -226,6 +229,71 @@ where
...
+    async fn create_sink(
+        &self,
+        request: Request<CreateSinkRequest>,
+    ) -> Result<Response<CreateSinkResponse>, Status> {
+        let mut sink = request.into_inner().sink.unwrap();
+
+        let id = self
+            .env
+            .id_gen_manager()
+            .generate::<{ IdCategory::Table }>()
```
Comment (neverchanje) [inline src/meta/src/rpc/service/ddl_service.rs:241]:
```suggestion
            .generate::<{ IdCategory::Sink }>()
```

File: src/sink/src/manager.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    fn drop_sink(&self, sink_id: &TableId) -> Result<()>;
+
+    /// Clear sinks, this is used when failover happens.
+    fn clear_sinks(&self) -> Result<()>;
+}
+
+/// `SinkColumnDesc` is used to describe a column in the Sink and is used as the column
+/// counterpart in `StreamScan`
+#[derive(Clone, Debug)]
+pub struct SinkColumnDesc {
```
Comment (neverchanje) [inline src/sink/src/manager.rs:40]: I think there's no such requirement to show the columns of a Sink? Users will only want to check the schema from the upstream materialized view.

File: src/sink/src/manager.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            data_type: c.data_type.clone(),
+            column_id: c.column_id,
+            skip_parse: false,
+        }
+    }
+}
+
+/// `SinkDesc` is used to describe a `Sink`
+#[derive(Clone, Debug)]
+pub struct SinkDesc {
```
Comment (neverchanje) [inline src/sink/src/manager.rs:60]: When will we use SinkDesc?

File: src/frontend/src/handler/show.rs [hunk 6]
```diff
@@ -84,7 +84,11 @@ pub fn handle_show_object(context: OptimizerContext, command: ShowObject) -> Res
             .iter_materialized_source()
             .map(|t| t.name.clone())
             .collect(),
-        ShowObject::Sink { _schema } => todo!(),
+        ShowObject::Sink { schema } => catalog_reader
+            .get_schema_by_name(session.database(), schema_or_default(&schema))?
+            .iter_source()
```
Comment (neverchanje) [inline src/frontend/src/handler/show.rs:89]:
```suggestion
            .iter_sink()
```

File: src/frontend/src/handler/create_sink.rs [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    let (schema_name, table_name) = Binder::resolve_table_name(stmt.sink_name.clone())?;
+    let (database_id, schema_id) = session
+        .env()
+        .catalog_reader()
+        .read_guard()
+        .check_relation_name_duplicated(
+            session.database(),
+            &schema_name,
+            &stmt.sink_name.to_string(),
+        )?;
```
Comment (neverchanje) [inline src/frontend/src/handler/create_sink.rs:86]: You already checked in `make_prost_sink`.

File: src/frontend/src/handler/create_sink.rs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+use crate::optimizer::PlanRef;
+use crate::planner::Planner;
+use crate::session::{OptimizerContext, OptimizerContextRef, SessionImpl};
+use crate::stream_fragmenter::StreamFragmenter;
+
+pub(crate) fn make_prost_sink(
+    session: &SessionImpl,
+    name: ObjectName,
+    materialized_view: ObjectName,
+    properties: HashMap<String, String>,
```
Comment (neverchanje) [inline src/frontend/src/handler/create_sink.rs:38]:
```suggestion
    database_id: DatabaseId,
    schema_id: SchemaId,
    sink_name: String,
    mv_name: String,
```

File: src/frontend/src/handler/drop_sink.rs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    #[tokio::test]
+    async fn test_drop_sink_handler() {
+        let sql_create_table = "create table t (v1 smallint);";
+        let sql_create_mv = "create materialized view mv as select v1 from t;";
+        let sql_create_sink = "create sink snk from mv with('sink' = 'mysql')";
+        let sql_drop_sink = "drop sink snk;";
+        let frontend = LocalFrontend::new(Default::default()).await;
+        frontend.run_sql(sql_create_table).await.unwrap();
+        frontend.run_sql(sql_create_mv).await.unwrap();
+        // frontend.run_sql(sql_create_sink).await.unwrap();
```
Comment (neverchanje) [inline src/frontend/src/handler/drop_sink.rs:60]:
```suggestion
        frontend.run_sql(sql_create_sink).await.unwrap();
```

I tried uncommenting this line and the test passed.

File: src/sink/src/manager.rs (hunk 5 above)
Comment (nanderstabel) [reply src/sink/src/manager.rs:60]: I peeked from the MemSourceManager implementation for inspiration not realizing column and sink description are not needed. I removed them.

File: src/frontend/src/catalog/schema_catalog.rs [hunk 10]
```diff
@@ -34,6 +37,10 @@ pub struct SchemaCatalog {
     source_by_name: HashMap<String, SourceCatalog>,
     source_name_by_id: HashMap<SourceId, String>,
 
+    // TODO(nanderstabel): Can be replaced with a bijective map: https://crates.io/crates/bimap.
```
Comment (neverchanje) [inline src/frontend/src/catalog/schema_catalog.rs:40]: 👍

File: src/frontend/src/handler/show.rs [hunk 11]
```diff
@@ -84,7 +84,7 @@ pub fn handle_show_object(context: OptimizerContext, command: ShowObject) -> Res
             .iter_materialized_source()
             .map(|t| t.name.clone())
             .collect(),
-        ShowObject::Sink { _schema } => todo!(),
+        ShowObject::Sink { schema: _ } => todo!(),
```
Comment (neverchanje) [inline src/frontend/src/handler/show.rs:87]: Returning an unimplemented error is preferred.

**Review by neverchanje — APPROVED** [review]
Generally lgtm. Looking forward the meta side impl :)

File: proto/catalog.proto [hunk 12]
```diff
@@ -33,6 +33,16 @@ message Source {
   string owner = 7;
 }
 
+message Sink {
+  uint32 id = 1;
+  uint32 schema_id = 2;
+  uint32 database_id = 3;
+  string name = 4;
+  string mv_name = 5;
```
Comment (yezizp2012) [inline proto/catalog.proto:41]:
```suggestion
  uint32 associated_table_id = 5;
```

Comment (nanderstabel) [reply proto/catalog.proto:41]: > 

Thank you for this suggestion, I will implement it in my next PR :)

**Pull request merged**